### PR TITLE
External configuration and image generation hook

### PR DESCRIPTION
**Pull request by hibatos**
Hi,
I wanted a way to configure the epaper station without wanting to either change my command modifying the station.py, so I implemented a configuration file where all the configuration defaults are stored.  
Overwriting the defaults with environment variables is still possible, so it is possible to update any existing setup to this version.

# Move config
For the configuration file I choose JSON, since it is easily extendable and parseable natively by most programming and scripting languages, so configuration changes could be easily automated.

- added config.json
- added json lib to station.py
- rewrote all configuration variables to adapt the new defaults source

## Automatic creation of work- and imgdir
To more easily deploy the station, I also added a feature that would allow the automated creation of the working and image directory, but I decided to keep this feature disabled by default.

It is configurable by setting the config options `createWorkdirIfNotExists` and `createImagedirIfNotExists` respectively to true.

# Periodic command execution for image generation/fetching
Also I needed a feature to periodically execute a command to fetch new images.
So I implemented a feature to allow setting up a command to be executed every x milliseconds.
To configure this to be active, you need to set the following configuration values:
- `imggenCommand` to a non zero-length command that shall get executed
- `imggenIntervalMs` to a non zero count of milliseconds for the interval of the command execution

For implementing this, I added the `subprocess` library  and added an interval timer, that is only active if `imggenIntervalMs > 0` and `len(imggenCommand) > 1` .

For the execution to be non-blocking/async, I decided to go with a thread, that gets executed by an interval timer.

# Graceful exit on coordinator stick malfunction/misconfiguration
Currently, if you remove the stick whilst the station is running or try to start the station whilst no stick is present, the station will crash.
This is no big deal, but I wanted to make it exit gracefully on failure, so I caught the init call errors, any `timaccop.run()` exceptions and made the station exit with `exit()` (more on that later).

# Refactoring and formatting
To not copy code, I moved the exit logic to a function `exit(exitCode:int)`, since I was going to need it for the graceful exit on failure above anyways.
Then - more for aesthetic reasons than anything else - I sorted the imports by name and function.

This of course are a lot of changes for a single commit, and if some changes are not needed/wanted/useful, I'd be willing to refactor this commit into many smaller ones, if need be.

And thanks for this awesome project. 

Best Regards
@hibatos

### CODE REVIEW

**Comment by danielkucera** [comment]
Hi @hibatos , nice work, thank you.
What do you think about changing the config from json to yaml? Ti would allow comments, save some quotes and be more readable IMO.

**Comment by hibatos** [comment]
I would agree, I am a bit biased, because I normally only work with and around Javascript and Typescript, both of which mainly use json.
The commentability would be rather nice though, if only for explanatory purposes.

**Comment by danielkucera** [comment]
Can you also please format the all integers as integers and hex strings as [ 0x00, 0x01, 0x02, ...] in the config?

**Comment by hibatos** [comment]
absolutely, I was not sure wether it would be favorable to keep in the same format as the env vars or a native format. I'll try to look into this tomorrow and wrap things up.

**Comment by danielkucera** [comment]
Sure, no problem, I will also consolidate the parsing of env vars afterwards, it is a mess I did not notice before.

**Comment by hibatos** [comment]
And one last thing, would it be favorable to rename the config keys to match the names of the environment variables or shall I keep the more descriptive names (things like the unit suffix for milliseconds or seconds)?

**Comment by danielkucera** [comment]
More descriptive is better, there is no documentation so it is better if they can be understood easily.

**Comment by hibatos** [comment]
I created the YAML file, any suggestions?

```yaml
%YAML 1.2
---
Configuration:
  # Storage setup
  Storage:
    Image directory: # Location where tag images are stored
      path: "./" # Relative or absolute path
      create if missing: false # Create the folder if it does not exist 
    Working directory: # Location where temporary files are stored
      path: "/tmp/" # Relative or absolute path 
      create if missing: false # Create the folder if it does not exist

  # Options for the tags 
  Tag setup:
    checkin interval in ms: 300000 # Interval in milliseconds the tags ask for new images
    retry interval in ms: 600000 # Interval in milliseconds the tags should retry to connect to the station
    failed checkins till blank: 100 # Checkin retry count after which the tags will go blank
    failed checkins till disassociate: 200 # Checkin retry count after which the tags will reeenter associate mode

  # Configuration for the zigbee stick, zigbee network and encryption
  Zigbee configuration:
    channel: 11 # Zigbee channel, (11-26 in the 2.4GHz band) choose whichever you achieve best signal stability with
    pan ID: [0x47, 0x44] # 16 bit network id, all devices in the same network should use the same
    extended address: [0x00, 0x12, 0x4B, 0x00, 0x14, 0xD9, 0x49, 0x35] # 64 bit MAC address for the station
    coordinator interface: "/dev/ttyACM1" # Serial interface of the zigbee stick (COM port)
    master key: "D306D9348E29E5E358BF2934812002C1" # Master key for the zigbee coordinator, 32 hex char (16 byte) long

  # Command that gets executed every x milliseconds whilst the station is running.
  # Will only run if command is not empty and interval > 0.
  External image generation command: 
    command: "" # Command to execute
    interval in ms: 0 # Interval of execution

```

**Comment by hibatos** [comment]
I just pushed the new implementation, I moved from JSON to YAML, removed the JSON file, added a config mapping to consolidate the environment variable parsing and moved all configuration variables into one dictionary.
For details check commit [c47518d](https://github.com/danielkucera/epaper-station/pull/23/commits/c47518dbf0a07caea9b13911b99f8cb48a18066e).

**Comment by hibatos** [comment]
And also I am currently doubting wether the external image generation process was a good idea, I think that a Cron job would be better suited for this. What do you think?

**Comment by danielkucera** [comment]
I would use camelCase for naming the variables, no spaces.
Regarding the scheduler, I personally use cron for generating images. It appears cleaner to me than pushing it to the daemon.

File: station.py [hunk 1]
```diff
@@ -129,77 +159,145 @@
...
+
+# get element from dict by path
+def get_element_by_path(object, keys):
+    if keys:
+        return get_element_by_path(object[keys[0]], keys[1:])
+    return object
+
+
+# generate config object from config mapping and config file
+def config_generator(config_mapping, config_file):
```
Comment (danielkucera) [inline station.py]: this should not be needed, the parser should output a sensible config object without any post-processing

**Comment by hibatos** [comment]
I'll remove the scheduling feature then and consolidate the configuration loader into one function.

**Comment by hibatos** [comment]
I merged the config functions, moved from arrow paths to dot paths and from spaced named to camel case in the configuration, removed the external image generation code and added the pyyaml library to the requirements.txt.

On that note, I also created a .gitignore and added ``__PYCACHE__`` to it.

**Comment by danielkucera** [comment]
Very nice. Just one nitpick: can you please rename `coordinatorInterface` to `interfacePort` . It is not a coordinator in fact.

**Comment by hibatos** [comment]
Oh, of course, never noticed that :)

**Comment by danielkucera** [comment]
Thank you.

**Pull request merged**